proposals/per-extension-language-preferences.md

```diff
@@ -0,0 +1,234 @@
+- disable then re-enable the extension.
+- upgrade the extension to a new version.
+
+#### Behavior of `i18n.getAllLanguages()`
+
+This method returns an array of language tags that the extension supported, e.g. `["en", "fr", "ja", "zh-CN"]`.
+This is a convenient method. Without this method, developers need to hardcode it into their code.
+
+#### Behavior of `i18n.onLanguageChanged` event
+
+When the user preferred language is changed to a new different language by `i18n.setCurrentLanguage(code)`, this event is fired.
+Developers use this event to know the language changed, and might do the following:
+
+- Update the content of extension pages in place (without refreshing the page)
+- Reload extension pages:
+  - If the page is stateless, refresh the page directly.
+  - If the page is stateful, save the current state and then refresh the page.
+  - Prompts the user that the language setting has changed, and asks the user if they want to refresh the page immediately.
+- Re-create the extension context menus, since the title of the context menus shoule be updated.
+- Update the action title and badge text if needed.
+
+#### Behavior of `i18n.getMessage()`, manifest and CSS files
+
+When this proposal is implemented, the behavior of these existing functions will change as follows:
+
+- If the preferred language is not set by `i18n.setCurrentLanguage()`, it is consistent with the existing behavior.
+- If the preferred language is set by `i18n.setCurrentLanguage()`, and the extension supports this language, then prioritize using that language.
+- If the preferred language is set by `i18n.setCurrentLanguage()`, but the extension doesn't support this language, then treat as if the preferred language is not set. This is an edge case, for example, the language was removed when the extension was upgraded.
+
+### Browser built-in language selection menu
+
+Based on the capabilities provided by this proposal, browsers could provide a unified built-in language selection menu for extensions, like Android and iOS.
+Whether to provide the built-in language selection menu and how to implement it is up to the browser to decide. This is just a suggestion.
+
+The built-in language selection menu has the following benefits:
+- Easy for developers: Developers do not need to implement it themselves, they only need to adapt to this proposal.
+- Easy for users: A unified UI makes it easier for users to use, otherwise each extension might be set up differently.
+- It is convenient for developers to develop and test i18n functions by switching the extension language.
+
+### Integrate other i18n implementations (frameworks) with browser.i18n
+
+This proposal can also be integrated with third-party i18n frameworks, as these frameworks typically allow developers to specify the language to be used.
+For example, developers use `i18n.getCurrentLanguage()` and `i18n.setCurrentLanguage()` for the user preferred language, but don't use `i18n.getMessage()`.
```

Comment:
No, they still need to add messages.json files for languages they support. Because the extension name, short name and description can only be localized with `browser.i18n` mechanism.

Reply:
@hanguokai @oliverdunk How about instead of specifying `builtin_languages_menu`, we add a `available_locales` property in the manifest listing all the locales supported by the extension. This would prevent the need for specifying the messages.json as sometimes (from my experience with some extensions) a language is supported to some extend without having messages present in messages.json. To overcome this one would need to copy the english messages.json but this seems very counter-intuitive. Specifying an empty messages.json is not sufficient to at least tell the Chrome Web Store a specific language is supported.